### ISSUE TITLE
[Mac] Unable to modify speech rate on MacOS

### ISSUE THREAD

**Issue by khanshoaib3** [stardew-access/stardew-access#410]
Changing the speech rate from the config doesn't work.
- by patrcis (on discord)

**Comment by OlzKennett** [comment]
I'm also having this issue. I changed the voice in accessibility in the read and speak setting in accessibility, and changed the speech rate in the config file from 220.0 to 400.0, but there is no change. Really hoping there is a fix for this. It's an epic mod! The draggy voice is the only thing getting in the way now.

**Comment by Flameborn** [comment]
I've been meaning to raise a discussion topic about the state of the MacOS speech wrapper for the mod, for various reasons.

Apple's two last speech synthesizer frameworks are NSSpeechSynthesizer from AppKit and AVSpeechSynthesizer from AVFoundation, these are specifically for synthesizing text.

When I wrote the speech wrapper, I only had access to an Intel Mac, which had a working NSSpeechSynthesizer framework. Even though it was already known that Apple will eventually phase it out, I've decided to go ahead and write an NSSpeechSynthesizer wrapper anyway, because:

* AVSpeechSynthesizer was still new, some features were not ironed out, e.g. it had issues with the default voice selection.
* There were still a lot of Intel Mac users and I wanted something that was both Intel and Silicon compatible.

As these were ironed out in later releases of MacOS, NSSpeechSynthesizer became less reliable and Apple now recommends AVSpeechSynthesizer. This was the cause for speech queuing suddenly breaking from one MacOS release to the next. It is, very likely, the cause for the speech rate suddenly not working as well.

These issues can be fixed if we rewrite the speech wrapper, however, I can't guarantee that it will work on Intel Macs as expected. This means that as far as the speech is concerned, new releases of the mod will be for Silicon only.

**Comment by OlzKennett** [comment]
This is all good to know. Thank you very much.

I’m assuming this won’t be until 1.8+ in that case? 

I’ve also got parallels so can run the windows version but am finding some issues with NVDA and it getting over excited and blurting out all movement such as dwasdawsdawsdaws… etc. I”m not really enough of a windows or NVDA user to hunt down the issue.

**Comment by Flameborn** [comment]
No problem.

If everyone's okay with dropping Intel Macs and depending on how much time I can make for this in the near future, it should be part of the next upcoming release, yes.

Regarding your NVDA issue, I think that might be because you have word echo, turned on.

**Comment by BrailleBennett** [comment]
Have you considered just using a pre-made library that supports all the things you want like:
https://github.com/m1maker/sral
That particular library would also close issue #459

**Comment by Flameborn** [comment]
From what I can tell, this library’s using the AVSpeechSynthesizer framework, relying on the isSpeaking method and the internal queue to queue speech. On newer MacOS versions, a runloop needs to be managed, similarly to the NSSpeechSynthesizer framework on earlier Macs. In this case, it would not solve much unfortunately.

In general, switching to this library could be beneficial for all platforms, but the AVSpeechSynthesizer engine support might need more work.

**Comment by BrailleBennett** [comment]
Would it be worth you submitting improvements to this library instead of writing it yourself on a new implementation for this mod only?
> From what I can tell, this library’s using the AVSpeechSynthesizer framework, relying on the isSpeaking method and the internal queue to queue speech. On newer MacOS versions, a runloop needs to be managed, similarly to the NSSpeechSynthesizer framework on earlier Macs. In this case, it would not solve much unfortunately.
> 
> 
> 
> In general, switching to this library could be beneficial for all platforms, but the AVSpeechSynthesizer engine support might need more work.

**Comment by Flameborn** [comment]
> Would it be worth you submitting improvements to this library instead of writing it yourself on a new implementation for this mod only?

Certainly, it would be more beneficial for everyone who uses this library. It is, however, generally more time consuming to write something which is part of an already defined system> you have to adhere to existing function calls, pass the expected return values, etc. For now, since time is not something I have lots of, I would rather spend it on ensuring that speech works for SDV, unless the switch to this library has been decided, in which case it might take more time. If we don't want to switch to this library in the near future then I think a native c# implementation of the wrapper would work better. There are lots of quirks with AVSpeechSynthesizer still, so I might have to return to the initial shared library implementation, but we'll see.

I am open to suggestions
